app/src/main/java/com/github/geohunt/app/data/repository/bounties/TeamsRepository.kt

```diff
@@ -0,0 +1,137 @@
+package com.github.geohunt.app.data.repository.bounties
+
+import com.github.geohunt.app.data.exceptions.TeamNotFoundException
+import com.github.geohunt.app.data.repository.UserRepositoryInterface
+import com.github.geohunt.app.model.Team
+import com.github.geohunt.app.utility.toMap
+import com.google.firebase.database.*
+import com.google.firebase.database.ktx.snapshots
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.tasks.await
+import kotlinx.coroutines.withContext
+
+/**
+ * Team repository for a specific bounty.
+ */
+class TeamsRepository(
+    bountyReference: DatabaseReference,
+    private val userRepository: UserRepositoryInterface,
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
+): TeamsRepositoryInterface {
+    private val teams = bountyReference.child("teams")
+
+    override suspend fun joinTeam(teamId: String, userId: String) {
+        withContext(ioDispatcher) {
+            if (getUserTeam(userId).first() != null) {
+                return@withContext
+            }
+
+            teams.child(teamId).child("members")
+                    .child(userId)
+                    .setValue(true)
+                    .await()
+        }
+    }
+
+
+    override suspend fun leaveTeam(userId: String) {
+        val team = getUserTeam(userId).first() ?: return
+
+        withContext(ioDispatcher) {
+            teams.child(team.teamId).child("members").child(userId).removeValue()
+        }
+    }
+
+    override fun getTeam(teamId: String): Flow<Team> {
+        return teams.child(teamId).snapshots
+                .map { snapshotToTeam(it) }
+                .flowOn(ioDispatcher)
+    }
+
+    override suspend fun getUserTeam(): Flow<Team?> = getUserTeam(userRepository.getCurrentUser().id)
```

Comment:
Shouldn't functions either return Flows or be suspend ?

Reply:
Again, very nice comments Alban, it's with great pleasure that I did not absolutely discard your comment :)

(note to the graders: Alban is my friend, I am just teasing him, he's currently very angry at me and threatened me to kill my mother, i know, this is quite unusual and I should delete him of my friend list on instagram so that he stops sending me his shitty posts)

Reply:
???